server/jetstream_cluster_3_test.go

```diff
@@ -106,3 +107,55 @@ func TestJetStreamClusterRemovePeerByID(t *testing.T) {
 	require_True(t, resp.Error == nil)
 	require_True(t, resp.Success)
 }
+
+func TestJetStreamClusterDiscardNewAndMaxMsgsPerSubject(t *testing.T) {
+	c := createJetStreamClusterExplicit(t, "R3S", 3)
+	defer c.shutdown()
+
+	// Client for API requests.
+	s := c.randomNonLeader()
+	nc, js := jsClientConnect(t, s)
+	defer nc.Close()
+
+	for _, test := range []struct {
+		name     string
+		storage  StorageType
+		replicas int
+	}{
+		{"MEM-R1", MemoryStorage, 1},
+		{"FILE-R1", FileStorage, 1},
+		{"MEM-R3", MemoryStorage, 3},
+		{"FILE-R3", FileStorage, 3},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			js.DeleteStream("KV")
+			// Make sure setting new without DiscardPolicy also being new is error.
+			cfg := &StreamConfig{
+				Name:          "KV",
+				Subjects:      []string{"KV.>"},
+				Storage:       test.storage,
+				MaxMsgsPer:    1,
+				AllowDirect:   true,
+				DiscardNewPer: true,
+				MaxMsgs:       10,
+				Replicas:      test.replicas,
+			}
+			if _, apiErr := addStreamWithError(t, nc, cfg); apiErr == nil {
+				t.Fatalf("Expected API error but got none")
+			} else if apiErr.ErrCode != 10052 || !strings.Contains(apiErr.Description, "discard new per subject requires discard new policy") {
+				t.Fatalf("Got wrong error: %+v", apiErr)
+			}
+			// Set broad discard new policy to engage DiscardNewPer
+			cfg.Discard = DiscardNew
+			addStream(t, nc, cfg)
+
+			// We want to test that we reject new messages on a per subject basis if the
+			// max msgs per subject limit has been hit, even if other limits have not.
+			_, err := js.Publish("KV.foo", nil)
+			require_NoError(t, err)
+
+			_, err = js.Publish("KV.foo", nil)
+			require_Error(t, err)
```

Comment:
would probably check, to be safe, that it would reject on MaxMsgsPer > 1, and that the error is the one we exxpect to get.

Reply:
ok makes sense.

Reply:
Good, but I meant in addition to test on MaxMsgsPer=10 for eg so that it only rejects messages once we hit the right number but allowing the rest, and then that the error you get on publishing the 11th message is the actual error you expect.

Reply:
Oh you mean the lowest level add stream, ok.

Reply:
ok added.